hddm/models/hddm_rl_regression.py

```diff
@@ -0,0 +1,296 @@
+
+        self.model_descrs = []
+
+        for model in models:
+            if isinstance(model, dict):
+                try:
+                    model_str = model['model']
+                    link_func = model['link_func']
+                except KeyError:
+                    raise KeyError("HDDMrlRegressor requires a model specification either like {'model': 'v ~ 1 + C(your_variable)', 'link_func' lambda x: np.exp(x)} or just a model string")
+            else:
+                model_str = model
+                link_func = lambda x: x
+
+            separator = model_str.find('~')
+            assert separator != -1, 'No outcome variable specified.'
+            outcome = model_str[:separator].strip(' ')
+            model_stripped = model_str[(separator+1):]
+            covariates = dmatrix(model_stripped, data).design_info.column_names
+
+            # Build model descriptor
+            model_descr = {'outcome': outcome,
+                           'model': model_stripped,
+                           'params': ['{out}_{reg}'.format(out=outcome, reg=reg) for reg in covariates],
+                           'link_func': link_func
+            }
+            self.model_descrs.append(model_descr)
+
+            print("Adding these covariates:")
+            print(model_descr['params'])
+            if group_only_regressors:
+                group_only_nodes += model_descr['params']
+                kwargs['group_only_nodes'] = group_only_nodes
+            self.reg_outcomes.add(outcome)
+
+        #set wfpt_reg
+        self.wfpt_reg_class = deepcopy(wfpt_reg_like)
+
+        super(HDDMrlRegressor, self).__init__(data, **kwargs)
+
+        # Sanity checks
+        for model_descr in self.model_descrs:
+            for param in model_descr['params']:
+                assert len(self.depends[param]) == 0, "When using patsy, you can not use any model parameter in depends_on."
+
+    def __getstate__(self):
+        d = super(HDDMrlRegressor, self).__getstate__()
+        del d['wfpt_reg_class']
+        for model in d['model_descrs']:
+            if 'link_func' in model:
+                print("WARNING: Will not save custom link functions.")
+                del model['link_func']
+        return d
+
+    def __setstate__(self, d):
+        d['wfpt_reg_class'] = deepcopy(wfpt_reg_like)
+        print("WARNING: Custom link functions will not be loaded.")
+        for model in d['model_descrs']:
+            model['link_func'] = lambda x: x
+        super(HDDMrlRegressor, self).__setstate__(d)
+
+    def _create_wfpt_knode(self, knodes):
+        wfpt_parents = self._create_wfpt_parents_dict(knodes)
+        wfpt_parents['alpha'] = knodes['alpha_bottom']
+        return Knode(self.wfpt_reg_class, 'wfpt', observed=True,
+                     col_name=['split_by', 'feedback', 'response', 'rt', 'q_init'],
+                     reg_outcomes=self.reg_outcomes, **wfpt_parents)
+
```

Comment:
Here you add a new method called `_create_stochastic_knodes` with the code I posted.

Reply:
Adding the following here while also removing the creation of alpha in base.py results in an error. Note I also keep the original create_stochastic_knodes (starting at line 238 here).

 def _create_stochastic_knodes(self, include):
        knodes = super(HDDMrlRegressor, self)._create_stochastic_knodes(includes)
        if 'alpha' in include:            
            knodes.update(self._create_family_normal_normal_hnormal('alpha', value=0, g_tau=50**-2, std_std=10))
        return knodes


---------------------------------------------------------------------------
KeyError                                  Traceback (most recent call last)
<ipython-input-5-5bd3dbb5434f> in <module>
----> 1 m = hddm.HDDMrlRegressor(data,'a ~ neural_regressor',include='alpha')
      2 m.sample(50,burn=30,dbname='traces.db',db='pickle')
      3 m.print_stats()

~/anaconda/envs/py36/lib/python3.6/site-packages/hddm/models/hddm_rl_regression.py in __init__(self, data, models, group_only_regressors, keep_regressor_trace, **kwargs)
    206         self.wfpt_reg_class = deepcopy(wfpt_reg_like)
    207 
--> 208         super(HDDMrlRegressor, self).__init__(data, **kwargs)
    209 
    210         # Sanity checks

~/anaconda/envs/py36/lib/python3.6/site-packages/hddm/models/hddm_info.py in __init__(self, *args, **kwargs)
    111         self.is_informative = kwargs.pop('informative', True)
    112 
--> 113         super(HDDM, self).__init__(*args, **kwargs)
    114 
    115     def _create_stochastic_knodes(self, include):

~/anaconda/envs/py36/lib/python3.6/site-packages/hddm/models/base.py in __init__(self, data, bias, include, wiener_params, p_outlier, **kwargs)
    727         self.wfpt_class = hddm.likelihoods.generate_wfpt_stochastic_class(wp, cdf_range=self.cdf_range)
    728 
--> 729         super(HDDMBase, self).__init__(data, **kwargs)
    730 
    731     def __getstate__(self):

~/anaconda/envs/py36/lib/python3.6/site-packages/hddm/models/base.py in __init__(self, data, **kwargs)
     38         self.std_depends = kwargs.pop('std_depends', False)
     39 
---> 40         super(AccumulatorModel, self).__init__(data, **kwargs)
     41 
     42 

~/anaconda/envs/py36/lib/python3.6/site-packages/kabuki/hierarchical.py in __init__(self, data, is_group_model, depends_on, trace_subjs, plot_subjs, plot_var, group_only_nodes)
    346         self.db = None
    347 
--> 348         self._setup_model()
    349 
    350     def _setup_model(self):

~/anaconda/envs/py36/lib/python3.6/site-packages/kabuki/hierarchical.py in _setup_model(self)
    350     def _setup_model(self):
    351         # create knodes (does not build according pymc nodes)
--> 352         self.knodes = self.create_knodes()
    353 
    354         #add data to knodes

~/anaconda/envs/py36/lib/python3.6/site-packages/hddm/models/base.py in create_knodes(self)
    759     def create_knodes(self):
    760         knodes = self._create_stochastic_knodes(self.include)
--> 761         knodes['wfpt'] = self._create_wfpt_knode(knodes)
    762 
    763         return list(knodes.values())

~/anaconda/envs/py36/lib/python3.6/site-packages/hddm/models/hddm_rl_regression.py in _create_wfpt_knode(self, knodes)
    231     def _create_wfpt_knode(self, knodes):
    232         wfpt_parents = self._create_wfpt_parents_dict(knodes)
--> 233         wfpt_parents['alpha'] = knodes['alpha_bottom']
    234         return Knode(self.wfpt_reg_class, 'wfpt', observed=True,
    235                      col_name=['split_by', 'feedback', 'response', 'rt', 'q_init'],

KeyError: 'alpha_bottom'

Reply:
is `alpha` in includes?

Reply:
Yes

Reply:
Then I'm not sure, maybe try with a debugger to see what's happening.

Reply:
I'll try to figure out how to solve it. Related to that, why shouldn't the knode for alpha be created in  hddm_info.py? It doesn't seem to influence the other classes.

Reply:
Well we could and it's probably fine for this. These inheritance abstractions exist exactly for this purpose, though.